Add upgradeProfile method

mauritsvanrees opened the pull request:
Note: this is based on a few lines from the CMFQuickInstaller from Plone, plus some extra lines for which I had inspiration:
https://github.com/plone/Products.CMFPlone/blob/4.3.6/Products/CMFPlone/QuickInstallerTool.py#L85

I am briefly wondering if that has influence on the license, although it is really just code that is very similar to what is already in GenericSetup, like in manage_doUpgrades, just structured a bit differently.

Anyway, this adds an upgradeProfile method which is basically a programmatic alternative to the manage_doUpgrades method. Instead of an optional list of step ids, you pass an optional version to which you want to upgrade.

Goal is to make code like this from a pull request for plone.app.upgrade easier:
https://github.com/plone/plone.app.upgrade/commit/ecc62fc27910557673b2f61094cd0c4e97876eda
There I am now first checking if a profile was previously installed. If not, then I ask the Quick Installer to upgrade the product, which then itself checks the profile. With the difference between product and profile it is a bit of a hassle, and the new method would help here. And I wanted to be able to upgrade to a specific version.

tseaver commented:
ISTM that the handling for a "miss" on a non-`None` `dest` should be "do nothing", rather than "assume latest" (which is fine for `dest=None`).  Something like:

```python
steps = [ug['step'] for ug in upgrades]
if dest is not None and dest not in steps:
    # log a warning and bail.
```

mauritsvanrees commented:
Good point. I updated the code.

mauritsvanrees commented:
I have updated the pull request with a related fix. When the last applied upgrade step had a checker, the profile version was not updated.  This was already the case in the manage_doUpgrades in the ZMI and I initially took over that logic.  Now we no longer look at the checker of the last applied step when deciding whether to set the profile version.  The checker, if any is set, normally returns True before running the step (it can be applied), and False afterwards (it was already applied). I added a test for it.

tseaver commented:
Thank you!

mauritsvanrees commented:
Thank you for reviewing and merging.
Any chance of a new release?

tseaver commented:
@mauritsvanrees https://pypi.python.org/pypi/Products.GenericSetup/1.7.7

mauritsvanrees commented:
Thanks a lot!
